torchdata/nodes/samplers/multi_dataset_weighted_sampler.py

```diff
@@ -0,0 +1,190 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import copy
+from typing import Any, Dict, Mapping, Optional
+
+import torch
+from torchdata.nodes.base_node import BaseNode, T
+from torchdata.nodes.samplers.utils import StopCriteria
+
+
+class MultiDatasetWeightedSampler(BaseNode[T]):
+    """A node that samples from multiple datasets with weights."""
+
+    DATASET_NODE_STATES_KEY = "dataset_node_states"
+    GENERATOR_STATE_KEY = "g_state"
+    DATASETS_EXHAUSTED_KEY = "_datasets_exhausted"
+
+    def __init__(
+        self,
+        source_nodes: Mapping[str, BaseNode[T]],
+        weights: Dict[str, float],
+        stopping_criterion: str = StopCriteria.CYCLE_UNTIL_ALL_DATASETS_EXHAUSTED,
```

Comment:
nit: let's keep consistent in naming 
```suggestion
        stop_criterion: str = StopCriterion.CYCLE_UNTIL_ALL_DATASETS_EXHAUSTED,
```